src/validation.cpp

```diff
@@ -862,7 +863,62 @@ static bool WriteBlockToDisk(const CBlock &block, FlatFilePos &pos,
 
 Amount GetBlockSubsidy(uint32_t nBits,
                        const Consensus::Params &consensusParams) {
-    return SUBSIDY;
+    static constexpr int64_t SUBSIDY_INT = SUBSIDY / SATOSHI;
+    if (!consensusParams.enableDifficultyBasedSubsidy) {
+        return SUBSIDY;
+    }
+    // We define:
+    // R: Reward
+    // D: Difficulty
+    // T: Target
+    // Tm: Max Target
+    // a: Base subsidy
+
+    // In Lotus, the reward R = a*log(D)
+    // R = a*log(Tm / T) = a*log(Tm) - a*log(T)
+
+    // We get the target by decomposing nBits:
+    // T = mantissa * 2^(exp - 3)
+
+    // The log of that cancels out the 2^x
+    // log(T) = log(mantissa) + exp - 3
+
+    // We can approximate log(x) using Log2FixedPoint.
+    // log(x) = Log2FixedPoint(2^b * x) / 2^b, where b is the precision.
+    // We use 16 bits for the precision.
+    static constexpr int64_t PRECISION_BITS = 16;
+
+    // log of max target, rounded up. It actually is 223.99997798605273, we
+    // correct for this by calculating with mantissa + 1.
+    static constexpr int64_t LOG_MAX_TARGET = 224;
+    const int32_t exp = nBits >> 24;
+    const uint32_t mantissa = (nBits & 0x007fffff);
+
+    // Compute a*log(nMantissa)
+    // We re-interpret mantissa as fixed point number with 16 bits in the
+    // fractional part, this is like dividing mantissa by 2^16.
+    // We correct for this later.
+    const int64_t logMantissaFixedPoint =
+        Log2FixedPoint(mantissa, PRECISION_BITS);
+
+    // Multiply by SUBSIDY before going back from fixed point to keep the
+    // precision. Otherwise, we only increase in SUBSIDY steps.
+    const int64_t aLogMantissa =
+        (SUBSIDY_INT * logMantissaFixedPoint) >> PRECISION_BITS;
+
+    // Compute a*log(2^(8 * (exp - 3))), which can be simplified to
+    // a*(8 * (exp - 3)). We need to add PRECISION_BITS here because we
+    // implicitly divided mantissa by 2^16 before and correct for that here.
+    const int64_t aLogExponent = SUBSIDY_INT * (8 * (exp - 3) + PRECISION_BITS);
+    const int64_t aLogTarget = aLogMantissa + aLogExponent;
```

Comment:
Nominally this would be:

```
aLogTarget = ((aLogMantissa << precision ) + (8 * (exp - (3 << precision)))) >> precision
```
But you already shifted the precision out above.

Reply:
why the `(3 << precision)`?